gui/system/forms.py

```diff
@@ -2853,6 +2853,59 @@ class Meta:
         model = models.CertificateAuthority
 
 
+class CertificateAuthoritySignCSRForm(ModelForm):
+    cert_CSRs = forms.ChoiceField(
+        choices=[]
+    )
+
+    cert_name = forms.CharField(
+        label=models.Certificate._meta.get_field('cert_name').verbose_name,
+        required=True,
+        help_text=models.Certificate._meta.get_field('cert_name').help_text
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(CertificateAuthoritySignCSRForm, self).__init__(*args, **kwargs)
+        self.fields['cert_CSRs'].choices = [(choice.pk, choice.cert_name) for choice in models.Certificate.objects.filter(cert_CSR__isnull=False)]
+
+    def save(self):
+        cdata = self.cleaned_data
+        choice = cdata.get('cert_CSRs')
+        ca = models.Certificate.objects.get(pk=choice)
+        cert_info = crypto.load_certificate(crypto.FILETYPE_PEM, self.instance.cert_certificate)
+        PKey = crypto.load_privatekey(crypto.FILETYPE_PEM, self.instance.cert_privatekey)
+        csr = crypto.load_certificate_request(crypto.FILETYPE_PEM, ca.cert_CSR)
+
+        cert = crypto.X509()
+        cert.set_serial_number(12345)
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(3650)
+        cert.set_issuer(cert_info.get_subject())
+        cert.set_subject(csr.get_subject())
+        cert.set_pubkey(csr.get_pubkey())
+        cert.sign(PKey, self.instance.cert_digest_algorithm)
+
+        new_cert = crypto.dump_certificate(crypto.FILETYPE_PEM, cert)
+        new_PKey = crypto.dump_privatekey(crypto.FILETYPE_PEM, PKey)
+        new_csr = models.Certificate(
+            cert_type=models.CERT_TYPE_EXISTING,
```

Comment:
Shouldn't this be CERT_TYPE_INTERNAL ? Since it is a new certificate generated internally?

Reply:
When you put it that way, yeah. I had different logic, since it was using the existing CSR and Internal CA to generate new one